Fix new unit test race in expiration-mailer tests

First seen here: https://github.com/letsencrypt/boulder/runs/6121711860?check_suite_focus=true

```
==================
WARNING: DATA RACE
Read at 0x00c000234408 by goroutine 90:
  github.com/letsencrypt/boulder/mocks.(*mockMailerConn).SendMail()
      /boulder/mocks/mocks.go:591 +0x14e
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendNags()
      /boulder/cmd/expiration-mailer/main.go:166 +0x1782
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendToOneRegID()
      /boulder/cmd/expiration-mailer/main.go:291 +0x3cb
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts.func1()
      /boulder/cmd/expiration-mailer/main.go:231 +0x172
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts·dwrap·1()
      /boulder/cmd/expiration-mailer/main.go:237 +0x66

Previous write at 0x00c000234408 by goroutine 10:
  github.com/letsencrypt/boulder/mocks.(*mockMailerConn).SendMail()
      /boulder/mocks/mocks.go:591 +0x234
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendNags()
      /boulder/cmd/expiration-mailer/main.go:166 +0x1782
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendToOneRegID()
      /boulder/cmd/expiration-mailer/main.go:291 +0x3cb
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts.func1()
      /boulder/cmd/expiration-mailer/main.go:231 +0x172
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts·dwrap·1()
      /boulder/cmd/expiration-mailer/main.go:237 +0x66

Goroutine 90 (running) created at:
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts()
      /boulder/cmd/expiration-mailer/main.go:228 +0x170
  github.com/letsencrypt/boulder/cmd/expiration-mailer.TestProcessCertsParallel()
      /boulder/cmd/expiration-mailer/main_test.go:234 +0x176
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 10 (running) created at:
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts()
      /boulder/cmd/expiration-mailer/main.go:228 +0x170
  github.com/letsencrypt/boulder/cmd/expiration-mailer.TestProcessCertsParallel()
      /boulder/cmd/expiration-mailer/main_test.go:234 +0x176
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000302e40 by goroutine 90:
  runtime.growslice()
      /usr/local/go/src/runtime/slice.go:162 +0x0
  github.com/letsencrypt/boulder/mocks.(*mockMailerConn).SendMail()
      /boulder/mocks/mocks.go:591 +0x184
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendNags()
      /boulder/cmd/expiration-mailer/main.go:166 +0x1782
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendToOneRegID()
      /boulder/cmd/expiration-mailer/main.go:291 +0x3cb
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts.func1()
      /boulder/cmd/expiration-mailer/main.go:231 +0x172
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts·dwrap·1()
      /boulder/cmd/expiration-mailer/main.go:237 +0x66

Previous write at 0x00c000302e40 by goroutine 10:
  github.com/letsencrypt/boulder/mocks.(*mockMailerConn).SendMail()
      /boulder/mocks/mocks.go:591 +0x1bc
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendNags()
      /boulder/cmd/expiration-mailer/main.go:166 +0x1782
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).sendToOneRegID()
      /boulder/cmd/expiration-mailer/main.go:291 +0x3cb
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts.func1()
      /boulder/cmd/expiration-mailer/main.go:231 +0x172
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts·dwrap·1()
      /boulder/cmd/expiration-mailer/main.go:237 +0x66

Goroutine 90 (running) created at:
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts()
      /boulder/cmd/expiration-mailer/main.go:228 +0x170
  github.com/letsencrypt/boulder/cmd/expiration-mailer.TestProcessCertsParallel()
      /boulder/cmd/expiration-mailer/main_test.go:234 +0x176
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 10 (running) created at:
  github.com/letsencrypt/boulder/cmd/expiration-mailer.(*mailer).processCerts()
      /boulder/cmd/expiration-mailer/main.go:228 +0x170
  github.com/letsencrypt/boulder/cmd/expiration-mailer.TestProcessCertsParallel()
      /boulder/cmd/expiration-mailer/main_test.go:234 +0x176
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
--- FAIL: TestProcessCertsParallel (2.11s)
    testing.go:[115](https://github.com/letsencrypt/boulder/runs/6121711860?check_suite_focus=true#step:6:115)2: race detected during execution of test
FAIL
FAIL	github.com/letsencrypt/boulder/cmd/expiration-mailer	12.[175](https://github.com/letsencrypt/boulder/runs/6121711860?check_suite_focus=true#step:6:175)s
```

--- a/mocks/mocks.go
+++ b/mocks/mocks.go
@@ -10,6 +10,7 @@ import (
 	"io/ioutil"
 	"math/rand"
 	"net"
+	"sync"
 	"time"
 
 	"github.com/jmhodges/clock"
@@ -561,6 +562,7 @@ func (*PublisherClient) SubmitToSingleCTWithResult(_ context.Context, _ *pubpb.R
 
 // Mailer is a mock
 type Mailer struct {
+	sync.Mutex
 	Messages []MailerMessage
 }
 
@@ -582,11 +584,15 @@ type MailerMessage struct {
 
 // Clear removes any previously recorded messages
 func (m *Mailer) Clear() {
+	m.Lock()
+	defer m.Unlock()
 	m.Messages = nil
 }
 
 // SendMail is a mock
 func (m *mockMailerConn) SendMail(to []string, subject, msg string) error {
+	m.parent.Lock()
+	defer m.parent.Unlock()
 	for _, rcpt := range to {
 		m.parent.Messages = append(m.parent.Messages, MailerMessage{
 			To:      rcpt,

